README
WHAT IS IT?
-----------

Ruby/Password is a suite of password handling methods for Ruby. It supports
the manual entry of passwords from the keyboard in both buffered and
unbuffered modes, password strength checking, random password generation,
phonemic password generation (for easy memorisation by human-beings) and the
encryption of passwords.

The common CrackLib library is used to perform password strength checking.

From the CrackLib README:

  CrackLib makes literally hundreds of tests to determine whether you've
  chosen a bad password.

  * It tries to generate words from your username and GECOS entry and tries
    to match them against the password you've chosen.

  * It checks for simplistic patterns.

  * It then tries to reverse-engineer your password into a dictionary
    word, and searches for it in your dictionary.

  - after all that, it's PROBABLY a safe(-ish) password.  8-)


The target audience for this library is system administrators who need
to write Ruby programs that prompt for, generate, verify and encrypt
passwords.


THIS FORK
---------

This is a fork from the original software written by Ian Macdonald. I've (Albert Lash of Savonix / Docunext) forked it to try and package it as a Ruby 1.9.1 gem
for my own convenience. I made the version 0.15.4 to try and differentiate from Ian's tree.

INSTALLATION
------------

Please see the INSTALL file for details of how to install Ruby/Password.


USAGE
-----

Please see the RDoc documentation for details of how to use Ruby/Password.


LICENCE
-------

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

Please see the file COPYING for the terms of the licence.



CHANGES


0.15.4
------

Trying to make compatible with 1.9.1. - Albert

0.5.3
-----

No functional changes. The build no longer tries to use packer.h if available.
crack.h will always be sought and used.


0.5.2
-----

No functional changes. The build environment was modified to search for the
system dictionary in the additional location of
/var/cache/cracklib/cracklib_dict.pwd, which is where it is on Debian Linux.


0.5.1
-----

Password.get would throw an exception in the unlikely event that STDIN reached
EOF without any input. Thanks to Will Drewry <wad@google.com> for pointing
this out.

pwgen now supports a -v or --version switch.


0.5.0
-----

This is a major update. The user-visible changes between this version and the
previous, 0.4.1, are as follows:

A new example program, pwgen, has been added, complete with man page. Ted
Ts'o's C program of the same name inspired this Ruby version. The man page is
also adapted from his.

A new class method, Password.phonemic, generates phonemic passwords. This,
too, was inspired by Ted Ts'o's work on pwgen. As its second parameter, it can
take zero or more new constants, logically OR'ed together. Available constants
are Password::ONE_DIGIT and Password::ONE_CASE. The former specifies that the
generated password should contain at least one digit, whilst the latter
specifies that the password should contain at least one capital letter.

The old Password.random method has been renamed Password.urandom and replaced
by a new Password.random that does not depend on /dev/urandom, making it much
more portable. Password.urandom will return nil on platforms that do not
implement the /dev/urandom device.

Password.get will now detect whether STDIN is connected to a tty. If not, no
password prompt is displayed and no attempt will be made to manipulate
terminal echo.

The prompt parameter to Password.get and Password.getc must now be passed in
its entirety. Previously, the string ': ' would be automatically appended to
the user-supplied prompt, but this behaviour has been changed in order to
accommodate null prompts. The default string for the prompt parameter has been
changed accordingly.

Running password.rb directly will now result in a call to Password.phonemic
and the display of the resulting password.

The Password::BadDictionary exception has been renamed
Password::DictionaryError and made a subclass of StandardError instead of
RuntimeError.

The CryptError exception has been moved to Password::CryptError and is now a
subclass of StandardError instead of RuntimeError.

A new constant, PASSWD_CHARS, gives the list of characters from which
automatically generated passwords will be chosen. Note that Password.urandom
will use the additional characters '+' and '/'.

A new constant, SALT_CHARS, gives the list of characters valid as salt
characters when invoking Password#crypt.

Password.getc now returns an instance of Password, not String.

Password.random now returns a Password, not a String.

A Password::CryptError exception is now raised if the salt passed to
Password#crypt contains a bad character.

RDoc documentation has been added. The INSTALL file explains how to build it.

The RD documentation has been removed in favour of the RDoc documentation.


0.4.1
-----

Use of the rb_enable_super() function caused a warning in Ruby 1.8.x. This
has now been fixed.


0.4.0
-----

When a bad dictionary path is provided to Password#check, a
Password::BadDictionary exception is now raised.

Turn off Ruby buffering for Password.getc, as this resulted in the prompt not
being displayed when called by Ruby 1.8.


0.3.0
-----

The Crack class is gone and Password is now a direct subclass of String.  If
you were previously creating Crack objects and calling their methods, you can
now no longer do this (and there really never was a good reason to do that,
anyway).

As a result of this change, Password#check now raises a Password::WeakPassword
exception if the password is weak, not a Crack::WeakPassword exception.


0.2.1
-----

No new features. The only changes are portability enhancements to the build
system.


0.2.0
-----

Password#check no longer returns an empty string on success and a string
describing the nature of the password weakness on failure.

Instead, it returns true on success and raises a Crack::WeakPassword exception
on failure.


0.1.1
-----

Password.get now returns an instance of Password, not String.
Password.new now defaults to assigning a null string.


test/tc_password.rb
#!/usr/bin/ruby -w
#
# $Id: tc_password.rb,v 1.3 2004/04/12 08:50:06 ianmacd Exp $

$: << File.dirname(__FILE__) + "/.." << File.dirname(__FILE__) + "/../lib"

require 'test/unit'
require 'password'


TIMES = 1000
LENGTH = 32


class TC_PasswordTest < Test::Unit::TestCase

  def test_check
    # Check for a weak password.
    pw = Password.new( 'foo' )
    assert_raises( Password::WeakPassword ) { pw.check }

    # Check for a good password.
    pw = Password.new( 'G@7flAxg' )
    assert_nothing_raised { pw.check }

    # Check for an exception on bad dictionary path.
    assert_raises( Password::DictionaryError ) { pw.check( '/tmp/nodict' ) }
  end

  def test_phonemic
    TIMES.times do |t|
      pw = Password.phonemic( LENGTH )
      assert( pw.length == LENGTH, "bad length: #{pw.length}, not #{LENGTH}" )
    end
  end

  def test_phonemic_one_case
    TIMES.times do |t|
      pw = Password.phonemic( LENGTH, Password::ONE_CASE )
      assert( pw =~ /[A-Z]/, "#{pw} has no upper-case letter" )
      assert( pw.length == LENGTH, "bad length: #{pw.length}, not #{LENGTH}" )
    end
  end

  def test_phonemic_one_digit
    TIMES.times do |t|
      pw = Password.phonemic( LENGTH, Password::ONE_DIGIT )
      assert( pw =~ /[0-9]/, "#{pw} has no digit" )
      assert( pw.length == LENGTH, "bad length: #{pw.length}, not #{LENGTH}" )
    end
  end

  def test_phonemic_one_case_one_digit
    TIMES.times do |t|
      pw = Password.phonemic( LENGTH, Password::ONE_CASE |
				      Password::ONE_DIGIT )
      assert( pw =~ /[A-Z]/, "#{pw} has no upper-case letter" )
      assert( pw =~ /[0-9]/, "#{pw} has no digit" )
      assert( pw.length == LENGTH, "bad length: #{pw.length}, not #{LENGTH}" )
    end
  end

  def test_random
    TIMES.times do |t|
      pw = Password.random( LENGTH )
      assert( pw.length == LENGTH, "bad length: #{pw.length}, not #{LENGTH}" )
    end
  end

  def test_urandom
    TIMES.times do |t|
      pw = Password.urandom( LENGTH )
      assert( pw.length == LENGTH, "bad length: #{pw.length}, not #{LENGTH}" )
    end
  end

  def test_crypt
    pw = Password.random( LENGTH )
    assert_nothing_raised { pw.crypt( Password::DES ) }
    assert_nothing_raised { pw.crypt( Password::MD5 ) }
    assert_raises( Password::CryptError ) { pw.crypt( Password::DES, '@*' ) }
  end

  def test_null_stdin
    $stdin.reopen( File.new( '/dev/null' ) )
    assert_nothing_raised { Password.get }
  end

end


rbcrack.c
/* rbcrack.c - a Ruby interface to CrackLib
 * 
 * $Id: rbcrack.c,v 1.20 2006/03/02 19:41:44 ianmacd Exp $
 *
 * Version : 0.5.3
 * Author  : Ian Macdonald <ian@caliban.org>
 *
 * Copyright (C) 2002-2006 Ian Macdonald
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2, or (at your option)
 *   any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software Foundation,
 *   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <ruby.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <crack.h>

#include "rbcrack.h"


VALUE ePassword_DictionaryError;
VALUE ePassword_WeakPassword;


/* <b>check</b><em>(dict=nil)</em>
 *
 * This interfaces to LibCrack to check the strength of the password. If
 * _dict_ is given, it is the path to the CrackLib dictionary, minus the
 * file's extension. For example, if the dictionary is located at
 * <tt>/usr/lib/cracklib_dict.pwd</tt>, _dict_ would be
 * <tt>/usr/lib/cracklib_dict</tt>. If it is not given, the dictionary found
 * at build time will be used.
 *
 * If a path is given that does not lead to a legible dictionary, a
 * Password::DictionaryError exception is raised. On success, +true+ is
 * returned. On failure, a Password::WeakPassword exception is raised.
 */
static VALUE passwd_check(VALUE self, VALUE args)
{
    VALUE dict;
    char *objection;
    char *buffer;

    /* pop the one and only argument we may have been passed */
    dict = rb_ary_pop(args);

    if (dict == Qnil || strcmp(STR2CSTR(dict), "") == 0) {
	/* no argument passed, so use default location from rbcrack.h */
	dict = rb_str_new2(CRACK_DICT);
    } else {
	buffer = malloc(strlen(STR2CSTR(dict)) + 8);
	strcpy(buffer, STR2CSTR(dict));
	strcat(buffer, ".pwd");

	if (access(buffer, R_OK) != 0) {
	    free(buffer);
	    rb_raise(ePassword_DictionaryError, "%s", strerror(errno));
	}

	free(buffer);

    }

    /* perform check on password */
    objection = FascistCheck(STR2CSTR(self), STR2CSTR(dict));

    /* return true on success; raise an exception otherwise */
    if (objection) {
	rb_raise(ePassword_WeakPassword, "%s", objection);
    } else {
	return Qtrue;
    }

}

/* initialize this class */
void Init_crack()
{
    VALUE cPassword;

    /* define the Password class */
    cPassword = rb_define_class("Password", rb_cString);

    /* define the Password::DictionaryError exception */
    ePassword_DictionaryError =
	rb_define_class_under(cPassword, "DictionaryError",
			      rb_eStandardError);

    /* define the Password::WeakPassword exception */
    ePassword_WeakPassword =
	rb_define_class_under(cPassword, "WeakPassword",
			      rb_eStandardError);

    /* define the Password.check method */
    rb_define_method(cPassword, "check", passwd_check, -2);

    return;
}


extconf.rb
# extconf for Ruby/Password
#
# $Id: extconf.rb,v 1.13 2006/03/02 17:35:06 ianmacd Exp $

require 'mkmf'

search_dicts = %w(
  /usr/local/lib/pw_dict.pwd
  /usr/lib/pw_dict.pwd
  /opt/lib/pw_dict.pwd
  /usr/local/lib/cracklib_dict.pwd
  /usr/lib/cracklib_dict.pwd
  /opt/lib/cracklib_dict.pwd
  /var/cache/cracklib/cracklib_dict.pwd
)

if dict = with_config('crack-dict')
  search_dicts.unshift(dict)
end

crack_dict = nil

# find the crack dictionary
print "checking for cracklib dictionary... "

search_dicts.each do |dict|
  # create a header file pointing to the crack dictionary
  if File.exist?(dict)
    puts dict
    crack_dict = dict.sub(/\.pwd/, '')
    break
  end
end

if crack_dict.nil?
  puts "no\nCouldn't find a cracklib dictionary on this system"
  exit 1
end

hfile = File.new("rbcrack.h", 'w')
hfile.printf("#define CRACK_DICT \"%s\"\n", crack_dict)
hfile.close

have_header('crack.h') && have_library('crack', 'FascistCheck') or exit 1

create_makefile('crack')

File.open('Makefile', 'a') do |f|
  f.print <<EOF

extra-clean:	distclean
		-rm -rf rbcrack.h doc/

docs:
		-rdoc -x CVS rbcrack.c lib

test:		crack.so FORCE
		-cd test; ./tc_password.rb

FORCE:
EOF
end


example/example.rb
#!/usr/bin/ruby -w
#
# $Id: example.rb,v 1.7 2004/04/07 09:49:06 ianmacd Exp $
#
# Copyright (C) 2002-2004 Ian Macdonald
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2, or (at your option)
#   any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software Foundation,
#   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

require 'password'

def handle_password( pw )
  pw.check
  puts pw.crypt( `uname` == "Linux\n" ? Password::MD5 : Password::DES )
end

begin
  my_string = Password.get( "Password with get: " )
  handle_password( my_string )
rescue Password::WeakPassword => reason
  puts reason
  retry
end

begin
  my_string = Password.getc( "Password with getc: ", 'X' )
  handle_password( my_string )
rescue Password::WeakPassword => reason
  puts reason
  retry
end


lib/password.rb
# $Id: password.rb,v 1.24 2006/03/02 19:42:33 ianmacd Exp $
# 
# Version : 0.5.3
# Author  : Ian Macdonald <ian@caliban.org>
#
# Copyright (C) 2002-2006 Ian Macdonald
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2, or (at your option)
#   any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software Foundation,
#   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


require 'crack'
require 'termios'


# Ruby/Password is a collection of password handling routines for Ruby,
# including an interface to CrackLib for the purposes of testing password
# strength.
# 
#  require 'password'
#
#  # Define and check a password in code
#  pw = Password.new( "bigblackcat" )
#  pw.check
#
#  # Get and check a password from the keyboard
#  begin
#    password = Password.get( "New password: " )
#    password.check
#  rescue Password::WeakPassword => reason
#    puts reason
#    retry
#  end
#
#  # Automatically generate and encrypt a password
#  password = Password.phonemic( 12, Password:ONE_CASE | Password::ONE_DIGIT )
#  crypted = password.crypt
#
#
class Password < String

  # This exception class is raised if an error occurs during password
  # encryption when calling Password#crypt.
  #
  class CryptError < StandardError; end

  # This exception class is raised if a bad dictionary path is detected by
  # Password#check.
  #
  class DictionaryError < StandardError; end

  # This exception class is raised if a weak password is detected by
  # Password#check.
  #
  class WeakPassword < StandardError; end

  VERSION = '0.5.3'

  # DES algorithm
  # 
  DES = true

  # MD5 algorithm (see <em>crypt(3)</em> for more information)
  #
  MD5 = false
 
  # This flag is used in conjunction with Password.phonemic and states that a
  # password must include a digit.
  #
  ONE_DIGIT  =	1

  # This flag is used in conjunction with Password.phonemic and states that a
  # password must include a capital letter.
  #
  ONE_CASE    = 1 << 1

  # Characters that may appear in generated passwords. Password.urandom may
  # also use the characters + and /.
  #
  PASSWD_CHARS = '0123456789' +
		 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' +
		 'abcdefghijklmnopqrstuvwxyz'

  # Valid salt characters for use by Password#crypt.
  #
  SALT_CHARS   = '0123456789' +
		 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' +
		 'abcdefghijklmnopqrstuvwxyz' +
		 './'

  # :stopdoc:

  # phoneme flags
  #
  CONSONANT = 1
  VOWEL	    = 1 << 1
  DIPHTHONG = 1 << 2
  NOT_FIRST = 1 << 3  # indicates that a given phoneme may not occur first

  PHONEMES = {
    :a	=> VOWEL,
    :ae	=> VOWEL      | DIPHTHONG,
    :ah => VOWEL      | DIPHTHONG,
    :ai => VOWEL      | DIPHTHONG,
    :b	=> CONSONANT,
    :c	=> CONSONANT,
    :ch	=> CONSONANT  | DIPHTHONG,
    :d	=> CONSONANT,
    :e	=> VOWEL,
    :ee	=> VOWEL      | DIPHTHONG,
    :ei	=> VOWEL      | DIPHTHONG,
    :f	=> CONSONANT,
    :g	=> CONSONANT,
    :gh	=> CONSONANT  | DIPHTHONG | NOT_FIRST,
    :h	=> CONSONANT,
    :i	=> VOWEL,
    :ie	=> VOWEL      | DIPHTHONG,
    :j	=> CONSONANT,
    :k	=> CONSONANT,
    :l	=> CONSONANT,
    :m	=> CONSONANT,
    :n	=> CONSONANT,
    :ng	=> CONSONANT  | DIPHTHONG | NOT_FIRST,
    :o	=> VOWEL,
    :oh	=> VOWEL      | DIPHTHONG,
    :oo	=> VOWEL      | DIPHTHONG,
    :p	=> CONSONANT,
    :ph	=> CONSONANT  | DIPHTHONG,
    :qu	=> CONSONANT  | DIPHTHONG,
    :r	=> CONSONANT,
    :s	=> CONSONANT,
    :sh	=> CONSONANT  | DIPHTHONG,
    :t	=> CONSONANT,
    :th	=> CONSONANT  | DIPHTHONG,
    :u	=> VOWEL,
    :v	=> CONSONANT,
    :w	=> CONSONANT,
    :x	=> CONSONANT,
    :y	=> CONSONANT,
    :z	=> CONSONANT
  }

  # :startdoc:


  # Turn local terminal echo on or off. This method is used for securing the
  # display, so that a soon to be entered password will not be echoed to the
  # screen. It is also used for restoring the display afterwards.
  #
  # If _masked_ is +true+, the keyboard is put into unbuffered mode, allowing
  # the retrieval of characters one at a time. _masked_ has no effect when
  # _on_ is +false+. You are unlikely to need this method in the course of
  # normal operations.
  #
  def Password.echo(on=true, masked=false)
    term = Termios::getattr( $stdin )

    if on
      term.c_lflag |= ( Termios::ECHO | Termios::ICANON )
    else # off
      term.c_lflag &= ~Termios::ECHO
      term.c_lflag &= ~Termios::ICANON if masked
    end

    Termios::setattr( $stdin, Termios::TCSANOW, term )
  end


  # Get a password from _STDIN_, using buffered line input and displaying
  # _message_ as the prompt. No output will appear while the password is being
  # typed. Hitting <b>[Enter]</b> completes password entry. If _STDIN_ is not
  # connected to a tty, no prompt will be displayed.
  #
  def Password.get(message="Password: ")
    begin
      if $stdin.tty?
	Password.echo false
	print message if message
      end

      pw = Password.new( $stdin.gets || "" )
      pw.chomp!

    ensure
      if $stdin.tty?
	Password.echo true
	print "\n"
      end
    end
  end


  # Get a password from _STDIN_ in unbuffered mode, i.e. one key at a time.
  # _message_ will be displayed as the prompt and each key press with echo
  # _mask_ to the terminal. There is no need to hit <b>[Enter]</b> at the end.
  #
  def Password.getc(message="Password: ", mask='*')
    # Save current buffering mode
    buffering = $stdout.sync

    # Turn off buffering
    $stdout.sync = true

    begin
      Password.echo(false, true)
      print message if message
      pw = ""

      while ( char = $stdin.getc ) != 10 # break after [Enter]
	putc mask
	pw << char
      end

    ensure
      Password.echo true
      print "\n"
    end

    # Restore original buffering mode
    $stdout.sync = buffering

    Password.new( pw )
  end


  # :stopdoc:
  
  # Determine whether next character should be a vowel or consonant.
  #
  def Password.get_vowel_or_consonant
    rand( 2 ) == 1 ? VOWEL : CONSONANT
  end

  # :startdoc:


  # Generate a memorable password of _length_ characters, using phonemes that
  # a human-being can easily remember. _flags_ is one or more of
  # <em>Password::ONE_DIGIT</em> and <em>Password::ONE_CASE</em>, logically
  # OR'ed together. For example:
  #
  #  pw = Password.phonemic( 8, Password::ONE_DIGIT | Password::ONE_CASE )
  #
  # This would generate an eight character password, containing a digit and an
  # upper-case letter, such as <b>Ug2shoth</b>.
  #
  # This method was inspired by the
  # pwgen[http://sourceforge.net/projects/pwgen/] tool, written by Theodore
  # Ts'o.
  #
  # Generated passwords may contain any of the characters in
  # <em>Password::PASSWD_CHARS</em>.
  #
  def Password.phonemic(length=8, flags=nil)

    pw = nil
    ph_flags = flags

    loop do

      pw = ""

      # Separate the flags integer into an array of individual flags
      feature_flags = [ flags & ONE_DIGIT, flags & ONE_CASE ]

      prev = []
      first = true
      desired = Password.get_vowel_or_consonant

      # Get an Array of all of the phonemes
      phonemes = PHONEMES.keys.map { |ph| ph.to_s }
      nr_phonemes = phonemes.size

      while pw.length < length do

	# Get a random phoneme and its length
	phoneme = phonemes[ rand( nr_phonemes ) ]
	ph_len = phoneme.length

	# Get its flags as an Array
	ph_flags = PHONEMES[ phoneme.to_sym ]
	ph_flags = [ ph_flags & CONSONANT, ph_flags & VOWEL,
		     ph_flags & DIPHTHONG, ph_flags & NOT_FIRST ]

	# Filter on the basic type of the next phoneme
	next if ph_flags.include? desired

	# Handle the NOT_FIRST flag
	next if first and ph_flags.include? NOT_FIRST

	# Don't allow a VOWEL followed a vowel/diphthong pair
	next if prev.include? VOWEL and ph_flags.include? VOWEL and
		ph_flags.include? DIPHTHONG

	# Don't allow us to go longer than the desired length
	next if ph_len > length - pw.length

	# We've found a phoneme that meets our criteria
	pw << phoneme

	# Handle ONE_CASE
	if feature_flags.include? ONE_CASE

	  if (first or ph_flags.include? CONSONANT) and rand( 10 ) < 3
	    pw[-ph_len, 1] = pw[-ph_len, 1].upcase
	    feature_flags.delete ONE_CASE
	  end

	end

	# Is password already long enough?
	break if pw.length >= length

	# Handle ONE_DIGIT
	if feature_flags.include? ONE_DIGIT

	  if ! first and rand( 10 ) < 3
	    pw << ( rand( 10 ) + ?0 ).chr
	    feature_flags.delete ONE_DIGIT

	    first = true
	    prev = []
	    desired = Password.get_vowel_or_consonant
	    next
	  end

	end

	if desired == CONSONANT
	  desired = VOWEL
	elsif prev.include? VOWEL or ph_flags.include? DIPHTHONG or
	      rand(10) > 3
	  desired = CONSONANT
	else
	  desired = VOWEL
	end

	prev = ph_flags
	first = false
      end

      # Try again
      break unless feature_flags.include? ONE_CASE or
		   feature_flags.include? ONE_DIGIT

    end

    Password.new( pw )

  end


  # Generate a random password of _length_ characters. Unlike the
  # Password.phonemic method, no attempt will be made to generate a memorable
  # password. Generated passwords may contain any of the characters in
  # <em>Password::PASSWD_CHARS</em>.
  #
  #
  def Password.random(length=8)
    pw = ""
    nr_chars = PASSWD_CHARS.size

    length.times { pw << PASSWD_CHARS[ rand( nr_chars ) ] }

    Password.new( pw )
  end


  # An alternative to Password.random. It uses the <tt>/dev/urandom</tt>
  # device to generate passwords, returning +nil+ on systems that do not
  # implement the device. The passwords it generates may contain any of the
  # characters in <em>Password::PASSWD_CHARS</em>, plus the additional
  # characters + and /.
  #
  def Password.urandom(length=8)
    return nil unless File.chardev? '/dev/urandom'

    rand_data = nil
    File.open( "/dev/urandom" ) { |f| rand_data = f.read( length ) }

    # Base64 encode it
    Password.new( [ rand_data ].pack( 'm' )[ 0 .. length - 1 ] )
  end


  # Encrypt a password using _type_ encryption. _salt_, if supplied, will be
  # used to perturb the encryption algorithm and should be chosen from the
  # <em>Password::SALT_CHARS</em>. If no salt is given, a randomly generated
  # salt will be used.
  #
  def crypt(type=DES, salt='')

    unless ( salt.split( // ) - SALT_CHARS.split( // ) ).empty?
      raise CryptError, 'bad salt'
    end

    salt = Password.random( type ? 2 : 8 ) if salt.empty?

    # (Linux glibc2 interprets a salt prefix of '$1$' as a call to use MD5
    # instead of DES when calling crypt(3))
    salt = '$1$' + salt if type == MD5

    # Pass to crypt in class String (our parent class)
    crypt = super( salt )

    # Raise an exception if MD5 was wanted, but result is not recognisable
    if type == MD5 && crypt !~ /^\$1\$/
      raise CryptError, 'MD5 not implemented'
    end

    crypt
  end

end


# Display a phonemic password, if run directly.
#
if $0 == __FILE__
  puts Password.phonemic
end
